python/ray/workflow/step_executor.py

```diff
@@ -372,7 +397,18 @@ def _workflow_step_executor(
             # Execute sub-workflow. Pass down "outer_most_step_id".
             with workflow_context.fork_workflow_step_context(
                     outer_most_step_id=outer_most_step_id):
-                result = execute_workflow(persisted_output)
+                sub_workflow = persisted_output
+                if resources_equal(runtime_options.ray_options,
+                                   sub_workflow.data.step_options.ray_options):
+                    # The workflow step returned by the current step has
+                    # exactly same resource and environment constraints.
+                    # So we execute it inplace.
+                    # TODO(suquark): We still have recursive Python calls.
+                    # This would cause stack overflow if we have a really
+                    # deep recursive call. We should fix it later.
```

Comment:
This looks dangerous. Should we have a feature flag for this feature? For example, .options(allow_inplace=True) then it'll run with in place optimization.